README.md
# gds-clusterconfig-to-sumologic-sync

A github action which takes information from a `gds.clusterconfig.*` repository, and syncs details to a Sumo Logic lookup table.

The lookup table will make it possible to query services, and find github events related to them, giving the developers a full picture as to how code made it to production.

## Requirements

1. the following **secret** must be present in your repository, either as a repository or organization level secret.

   1. `SUMO_LOGIC_API_ACCESS_SECRET`

      ```json
      {
        "SUMOLOGIC_ACCESS_ID": "insert-id-here",
        "SUMOLOGIC_ACCESS_KEY": "insert-key-here",
        "SUMOLOGIC_API_ENDPOINT": "https://api.us2.sumologic.com/api"
      }
      ```

## Configuration

| Input            | Description                                                  | Default      |
| ---------------- | ------------------------------------------------------------ | ------------ |
| sumologic_config | See `SUMO_LOGIC_API_ACCESS_SECRET` in **Requirements**       | **REQUIRED** |
| cluster          | eg. p01, i01, etc...<br />This is **optional** if the repository name follows the `gds[.](china[.])?clusterconfig[.][jips][0-9]{2}` naming convention. | ""           |

## Example Usage

```yaml
name: Sync to Sumo Logic
on:
  push:
    branches:
      - main
jobs:
  sumologic:
    runs-on: ubuntu-latest
    name: Sync GDS clusterconfig with Sumo Logic
    steps:
    - name: Checkout code
      id: checkout
      uses: actions/checkout@v2
    - name: Sync to Sumo Logic
      id: sync
      uses: glg-public/gds-clusterconfig-to-sumologic-sync@main
      with:
        sumologic_config: ${{ secrets.SUMO_LOGIC_API_ACCESS_SECRET }}
```

## [Development](./DEVELOPMENT.md)


CHANGELOG.md
# Changelog

- 1.2.0
  - `entrypoint.sh` now passes the insert, update, delete off to the `publish-updates.js` script
  - rows that are no longer used are now being cleaned up in the lookup tables
  - created `Makefile` for development
  - added throttling to the requests to Sumo Logic to deal with rate limit issues
- 1.1.0
  - Using new tables
    - `Share Your Ideas/GDS/Tables/cc_cluster_service`
    - `Share Your Ideas/GDS/Tables/cc_cluster_service_repo_branch`
- 1.0.0
  - Initial release
  - Populates to the `Share Your Ideas/GDS/Lookups/cluster_service` table.


entrypoint.sh
#!/bin/bash -l
set -o nounset;
set -o errexit;
set -o pipefail;

IFS=$'\n\t'

function lookup_upload_entry_heredoc () {
  local OUTPUT_FILE="$1"
  cat << DOC |
{
  "cluster": "${CLUSTER}",
  "service": "${SERVICE}",
  "ecr_repo": "${ECR_REPO}",
  "ecr_tag": "${ECR_TAG}",
  "git_repo": "${GIT_REPO}",
  "git_branch": "${GIT_BRANCH}"
}
DOC
jq -cM '.' >> "${OUTPUT_FILE}"
}

if [[ "${GITHUB_REPOSITORY}" =~ [^/]+\/gds\.(china\.)?clusterconfig\.(.*) ]]; then
  # otherwise it has to match the gds clusterconfig repo name syntax
  CLUSTER="${BASH_REMATCH[2]}"
else
  # override if provided via the action
  CLUSTER="${INPUT_CLUSTER}"
  if [[ -z "${CLUSTER}" ]]; then
    echo "Your repository must be named gds.clusterconfig.* or you have to" \
      "provide the 'cluster' github action parameter"
          exit 1
  fi
fi
export CLUSTER

rm -rf '/tmp/payload'
while IFS= read -r -d '' FILE; do
  unset SERVICE ECR_REPO GIT_REPO GIT_BRANCH TYPE REPOSITORY
  # service is based on directory name
  SERVICE="$(basename "$(dirname "$FILE")")"

  # parse out the deploy commands we support
  IFS=$' \t' read -r TYPE REPOSITORY <<< \
    "$(grep -e "^\(autodeploy\|dockerdeploy\|dockerbuild\)\s" "$FILE" | tail -n1)"

  if [[ "${TYPE}" == "dockerdeploy" ]]; then
    # eg. github/glg/epi-screamer/gds-migration:latest
    if [[ "${REPOSITORY}" =~ ([^/]+)\/([^/]+)\/([^/]+)\/([^:]+)(:(.*))? ]]; then
      #                      ↑        ↑        ↑        ↑      ↑ ↑
      #                      |        |        |        |      5 6 docker tag
      #                      |        2 org    3 repo   4 branch
      #                      1 source control provider (eg. github)
      ECR_REPO="${REPOSITORY}"
      ECR_REPO="${BASH_REMATCH[1]}/${BASH_REMATCH[2]}/${BASH_REMATCH[3]}/${BASH_REMATCH[4]}"
      # ERE does not support non-capturing groups, that's why branch is in 6
      ECR_TAG="${BASH_REMATCH[6]:-latest}"
      GIT_REPO="${BASH_REMATCH[2]}/${BASH_REMATCH[3]}"
      GIT_BRANCH="${BASH_REMATCH[4]}"
    fi
  fi

  if [[ "${TYPE}" == "dockerbuild"  ]]; then
    # NOTE: dockerbuild WITHOUT branch specification won't work, but
    #       should also not be allowed via cc-screamer
    #       eg. git@github.com:glg/log.git#master
    if [[ "${REPOSITORY}" =~ git@github.com:([^#]+).git#(.*) ]]; then
      #                                     ↑           ↑
      #                                     1 org/repo  2 branch
      ECR_REPO="${BASH_REMATCH[1]}/${BASH_REMATCH[2]}"
      ECR_TAG="latest"
      GIT_REPO="${BASH_REMATCH[1]}"
      GIT_BRANCH="${BASH_REMATCH[2]}"
    fi
  fi

  if [[ "${TYPE}" == "autodeploy" ]]; then
    # NOTE: autodeploy WITHOUT branch specification won't work, but
    #       should also not be allowed via cc-screamer
    #       eg. git@github.com:glg/log.git#master
    if [[ "${REPOSITORY}" =~ git@github.com:([^#]+).git#(.*) ]]; then
      #                                     ↑           ↑
      #                                     1 org/repo  2 branch
      ECR_REPO="${CLUSTER}.glgresearch.com/${SERVICE}"
      ECR_TAG="latest"
      GIT_REPO="${BASH_REMATCH[1]}"
      GIT_BRANCH="${BASH_REMATCH[2]}"
    fi
  fi

  if [[ -z "${GIT_REPO:-}" ]]; then
    echo "warning: ${CLUSTER}/${FILE}: unable to extract GIT_REPO"
    continue
  fi

  lookup_upload_entry_heredoc '/tmp/payload'

done < <(find . -maxdepth 2 -type f -name orders -print0)

# need to make sure the file exists, even if the CC contains no orders
touch '/tmp/payload'
node '/app/uploader/process-updates.js'


uploader/sumologic.js
'use strict';
const axios = require('axios');

const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

const sumoRequest = async ({
  sumo, url, admin=false, method='get', payload=null, preSleep=0
}) => {
  // sometimes it's prefixed with the endpoint, sometimes not
  const safeUrl = url.replace(sumo.endpoint, '');
  const args = {
    method
    , url: `${sumo.endpoint}${safeUrl}`
    , auth: {
      username: sumo.id
      , password: sumo.key
    }
    , headers: {}
    , validateStatus: false
  }
  if (admin === true) {
    args.headers.isAdminMode = 'true';
  }
  if (payload !== null) {
    args.headers['Content-Type'] = 'application/json';
    args.data = payload;
  }
  await sleep(preSleep);
  const { status, data } = await axios(args);
  return { status, data };
};

const sumoJob = async ({sumo, url, admin=false, method='get'}) => {
  const initUrl = url;
  const statusUrl = id => `${url}/${id}/status`;
  const resultUrl = id => `${url}/${id}/result`;

  // queue the job with sumologic
  const { data: { id: jobId } } = await sumoRequest({
    sumo, url: initUrl, admin, method
  });
  console.log(`:: ${url} jobId ${jobId}`);

  // wait for job to complete
  const success = await [1500, 7500, 15000, 30000, 60000]
  .reduce(async (success, ms) => {
    success = await success; // process sequentially
    if (success) {
      return success; // exit if job is done
    }
    // wait before request (intentional sleep before first request)
    // with 1.5 seconds buffer, we run 40 jobs per minute max
    // 40 jobs * 3 min api calls = 120, which is still way under the 240/m
    // rate limit of sumo logic
    console.log(`:: ${url} check job in ${ms}ms`);
    await sleep(ms);
    const { status, data } = await sumoRequest({
      sumo, url: statusUrl(jobId), admin
    });
    if (status === 200 && data.status === 'Success') {
      return true;
    }
    console.log(`:: ${url} ${JSON.stringify({status, data}, null, 0)}`);
    return false
  }, false);

  if (!success) {
    throw new Error(`:: ${url} job failed`);
  }

  const { status, data } = await sumoRequest({
    sumo, url: resultUrl(jobId), admin
  });

  if (status !== 200) {
    throw new Error(`:: ${url} could not fetch data`);
  }

  return data;
};

// https://help.sumologic.com/APIs/Search-Job-API/About-the-Search-Job-API
//
// 301 moved
// 401 unauthorized
// 403 forbidden
// 404 notfound
// 405 method.unsupported
// 415 contenttype.invalid
// 429 rate.limit.exceeded
// 500 internal.error
// 503 service.unavailable
const sumoSearch = async ({sumo, url: initUrl, payload}) => {
  // queue the job with sumologic
  //
  // 400 generic
  // 400 invalid.timestamp.to
  // 400 invalid.timestamp.from
  // 400 to.smaller.than.from
  // 400 unknown.timezone
  // 400 empty.timezone
  // 400 no.query
  // 400 unknown.time.type
  // 400 parse.error
  //
  // error:
  // {
  //   "status" : 400,
  //   "id" : "IUUQI-DGH5I-TJ045",
  //   "code" : "searchjob.invalid.timestamp.from",
  //   "message" : "The 'from' field contains an invalid time."
  // }
  //
  // success:
  // {
  //   "status": 202,
  //   "data": {
  //     "id": "641259F2B975D432",
  //     "link": {
  //       "rel": "self",
  //       "href": "https://api.us2.sumologic.com/api/v1/search/jobs/641259F2B975D432"
  //     }
  //   }
  // }
  const res = await sumoRequest({
    sumo
    , url: initUrl
    , admin: false
    , method: 'post'
    , payload
  });
  if (!(res.status >= 200 && res.status <= 299)) {
    console.error(res);
    throw new Error(res.code ?? 'could not queue search');
  }
  const {data: {id: jobId, link: {href: statusUrl}}} = res;
  console.log(`:: jobId='${jobId}' statusUrl='${statusUrl}'`);

  // due to sumo logic's rate limits, we pre-emptively wait before checking status
  const success = await [1500, 7500, 15000, 30000, 60000]
  .reduce(async (success, ms) => {
    success = await success; // process sequentially
    if (success) {
      return success; // exit if job is done
    }
    // wait before request (intentional sleep before first request)
    // with 1.5 seconds buffer, we run 40 jobs per minute max
    // 40 jobs * 3 min api calls = 120, which is still way under the 240/m
    // rate limit of sumo logic
    console.log(`:: jobId='${jobId}' check job status in ${ms}ms`);
    await sleep(ms);

    // "NOT STARTED"
    // "GATHERING RESULTS"
    // "FORCE PAUSED"
    // "DONE GATHERING RESULTS"
    // "CANCELLED"
    // {
    //    "state":"DONE GATHERING RESULTS",
    //    "messageCount":90,
    //    "histogramBuckets":[
    //       {
    //          "length":60000,
    //          "count":1,
    //          "startTimestamp":1359404820000
    //       }
    //    ],
    //    "pendingErrors":[],
    //    "pendingWarnings":[],
    //    "recordCount":1
    // }
    const {status, data} = await sumoRequest({sumo, url: statusUrl});
    const {state, messageCount, recordCount} = data;

    if (status >= 200 && status <= 299 && state === "DONE GATHERING RESULTS") {
      console.log(`:: messageCount='${messageCount}' recordCount='${recordCount}'`);
      return true;
    }
    console.log(`:: ${url} ${JSON.stringify({status, data}, null, 0)}`);
    return false
  }, false);

  if (!success) {
    throw new Error(`:: ${url} job failed`);
  }

  const resultUrl = (jobId) => `${statusUrl}/messages?offset=0&limit=10000`
  const { status, data } = await sumoRequest({
    sumo, url: resultUrl(jobId)
  });

  if (status !== 200) {
    throw new Error(`:: ${url} could not fetch data`);
  }

  return data;
};

module.exports = {
  sumoRequest
  , sumoJob
  , sumoSearch
}


uploader/process-updates.js
'use strict';
const { mkdir, readFile, writeFile, access } = require('fs/promises');
const { constants: { F_OK } } = require('fs');
const { sumoRequest, sumoSearch } = require('./sumologic');

// NOTE: The "main()" method is at the bottom of the file, start there!
// https://api.us2.sumologic.com/docs/#tag/lookupManagement

const createConfig = () => {
  // GitHub Action Inputs
  let {
    SUMOLOGIC_ACCESS_ID: id
    , SUMOLOGIC_ACCESS_KEY: key
    , SUMOLOGIC_API_ENDPOINT: endpoint
  } = JSON.parse(process.env.INPUT_SUMOLOGIC_CONFIG ?? "{}");
  // Override if environment variables are present
  id = process.env.SUMOLOGIC_ACCESS_ID ?? id;
  key = process.env.SUMOLOGIC_ACCESS_KEY ?? key;
  endpoint = process.env.SUMOLOGIC_API_ENDPOINT ?? endpoint;
  // return config object
  return {
    sumo: { id, key, endpoint }
    , dataDir: '/tmp/sumologic_data'
    , clusterConfigExport: '/tmp/payload'
    , repoDir: process.env.GITHUB_REPOSITORY
    , targetCluster: process.env.CLUSTER
    // TODO: Need to make these configurable, but have to find proper place
    , tableIds: ["0000000001007719", "0000000000FF668A"] // PROD
    //, tableIds: ["00000000010D6C07", "00000000010D4453"] // DEV
  };
};

const cacheWrapper = async (cachePath, method) => {
  return access(cachePath, F_OK)
  .then(
    async () => {
      const data = await JSON.parse(await readFile(cachePath, 'utf8'));
      console.log(`:: read from cache ${cachePath}`);
      return data;
    }
    , async () => {
      const data = await method();
      if (data !== null) {
        const json = JSON.stringify(data, null, 2);
        await writeFile(cachePath, json, 'utf8');
      }
      return data;
    }
  );
};

const mapC = async (data, callback, {concurrency=1} = {}) => {
  const _data = Array.from(data);
  const result = new Array(_data.length);
  await Promise.all(
    new Array(concurrency).fill(_data.entries())
    .map(async iterator => {
      for (let [index, item] of iterator) {
        result[index] = await callback(item, index);
      }
    })
  );
  return result;
};

const kv = obj => Object.entries(obj).map(([k, v]) => `${k}='${v}'`).join(' ');

const getTableInfo = async ({sumo, dataDir, tableId}) => {
  const outputFile =  `${dataDir}/${tableId}.json`;
  const res = await cacheWrapper(
    outputFile
    , () => sumoRequest({
      sumo
      , url: `/v1/lookupTables/${tableId}`
      , admin: true
    })
  );
  const {status, data} = res;
  if (!(status >= 200 && status <= 299)) {
    throw new Error(JSON.stringify(res));
  }
  const {contentPath, name} = data;
  return {table: {name, path: contentPath, id: tableId}};
};

const createSearchPayload = ({path}) => {
  const query = `cat path://"${path}"`
  const from = (Date.now() - 1000).toString();
  const to = Date.now().toString();
  return JSON.stringify({query, from, to});
};

const fetchClusterLookups = async ({sumo, dataDir, repoDir, targetCluster, tableIds}) => {
  console.log(`:: processing ${tableIds.length} table(s)`);
  const result = {};
  await mapC(tableIds
    , async (tableId, index) => {
      console.log(`:: table_index='${index}' id='${tableId}'`);
      const {table} = await getTableInfo({sumo, dataDir, tableId});
      console.log(`:: ${kv(table)}`);

      const payload = createSearchPayload(table);

      const search = await cacheWrapper(
        `${dataDir}/search-${table.id}.json`
        , () => sumoSearch({
          sumo
          , url: `/v1/search/jobs`
          , payload
        })
      );
      result[table.id] = search.messages.reduce(
        (final, {map: {cluster, service, git_repo, git_branch, ecr_repo, ecr_tag}}) => {
          if (cluster === targetCluster) {
            final[`${cluster}|${service}|${git_repo}|${git_branch}`] = {
              cluster, service, git_repo, git_branch, ecr_repo, ecr_tag
            };
          }
          return final;
        }
        , {}
      );
      return result[table.id];
    }
    // We really want concurrency to currently be at 1, but it's built into
    // the code to scale, if sumo logic increases the rate limit for
    // APIs up from 240/min/user
    , {concurrency: 1}
  )
  return result;
};

const loadClusterServices = async ({clusterConfigExport}) => {
  const input = await readFile(clusterConfigExport, 'utf8');
  return input
  .trim()
  .split('\n')
  .filter(v => v) // remove empty rows from array
  .map(row => JSON.parse(row));
};

const createLookupPayload = (entry) => ({
  "row": Object.entries(entry).map(([k, v]) => ({ "columnName": k, "columnValue": v }))
});

const deleteLookupPayload = (entry) => ({
  "primaryKey": Object.entries(entry).map(([k, v]) => ({ "columnName": k, "columnValue": v }))
});

const uploadToLookups = async ({sumo}, clusterServices, lookupTable) => {
  // loop through all the cluster/service entries
  await mapC(clusterServices
    , async (entry, index) => {
      // process them once per lookup table
      for (const tableId of Object.keys(lookupTable)) {
        console.log(`:: start upload ${kv({tableId})} ${kv(entry)}`);
        // regardless of success, we'll remove any attempted processing on entries
        // from the list, so they are not removed.
        delete lookupTable[tableId][
          `${entry.cluster}|${entry.service}|${entry.git_repo}|${entry.git_branch}`
        ]
        // upload the entry to sumo logic. logs errors, but proceed regardless.
        const { status, data } = await sumoRequest({
          sumo
          , url: `/v1/lookupTables/${tableId}/row`
          , method: 'put'
          , payload: createLookupPayload(entry)
          , preSleep: 1500
        })
        .catch(error => {
          console.error(error);
          return {status: -1, data: null};
        });
        console.log(`:: end upload ${kv({tableId})} ${kv({status})} ${kv(entry)}`);
      }
    }
    , {concurrency: 1}
  );
};

const removeExpiredEntries = async({sumo}, lookupTable) => {
  for (const [tableId, entries] of Object.entries(lookupTable)) {
    for (const {cluster, service, git_repo, git_branch} of Object.values(entries)) {
      const primaryKeys = {cluster, service, git_repo, git_branch};
      console.log(`:: start delete ${kv({tableId})} ${kv(primaryKeys)}`);
      const { status, data } = await sumoRequest({
        sumo
        , url: `/v1/lookupTables/${tableId}/deleteTableRow`
        , method: 'put'
        , payload: deleteLookupPayload(primaryKeys)
        , preSleep: 1500
      })
      .catch(error => {
        return {status: -1, data: null};
      });
      console.log(`:: end delete ${kv({tableId})} ${kv({status})} ${kv(primaryKeys)}`);
    }
  }
};

(async () => {
  try {
    /*
     * MAIN
     */
    const config = createConfig();

    await mkdir(config.dataDir, { recursive: true });

    const clusterServices = await loadClusterServices(config);
    const lookupTable = await fetchClusterLookups(config);
    // NOTE: commenting out the upload would clear the lookup table data for this cluster
    await uploadToLookups(config, clusterServices, lookupTable);
    await removeExpiredEntries(config, lookupTable);

  } catch (error) {
    // important to exit with a failure code for the action to abort
    console.error(error);
    process.exit(1);
  }
})();
